bpf/lib/nodeport.h

```diff
@@ -1144,7 +1145,28 @@ static __always_inline bool snat_v4_needed(struct __ctx_buff *ctx, __be32 *addr,
 		if (identity_is_remote_node(info->sec_label))
 			return false;
 #endif
-#if defined(ENABLE_EGRESS_GATEWAY)
+
+		/* Check if this packet belongs to reply traffic coming from a
+		 * local endpoint.
+		 *
+		 * If ep is NULL, it means there's no endpoint running on the
+		 * node which matches the packet source IP, which means we can
+		 * skip the CT lookup since this cannot be reply traffic.
+		 */
+		if (ep) {
+			bool is_reply = false;
+			struct ipv4_ct_tuple tuple = {
+				.nexthdr = ip4->protocol,
+				.daddr = ip4->daddr,
+				.saddr = ip4->saddr
+			};
+
+			if (!ct_is_reply4(get_ct_map4(&tuple), ctx, ETH_HLEN + ipv4_hdrlen(ip4),
+					  &tuple, &is_reply) && is_reply)
+				ct_is_reply = true;
```

Comment:
Why can't we just `return false` here instead? This would eliminate the `if(!ct_is_reply)` below.

Reply:
No strong reason here, I just thought avoiding this early return would make the code below a bit more explicit

Reply:
I think it's in opposite - the early return improves the code readability (no need to keep the context in head until the end of the function), and reduces the instr count.